Mirantis/blueprint-operator
# [BOP-102] Create/Update/Delete manifest objects using dynamic client

JIRA: [BOP-102](https://mirantis.jira.com/browse/BOP-102)

For creating/updating/deleting resources for Manifest based addons, this PR uses dynamic objects ([unstructured](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured)) instead of concrete types (e.g. `Deployment`, `Service`, `CustomResourceDefinition` etc).

The biggest advantage of doing this is that we can now support any kind of Kubernetes resource, including custom resources.

In a future PR, our `HelmController` will also be installed using the same dynamic client code to install  

## Unit Tests

There are few unit tests for verifying the basic functionality of the dynamic client: create/update/delete single/multiple objects from the manifest file

## Manual Testing

### TEST SCENARIO 1: Install add-on metallb with version v0.13.10
1. Create manifest addons: 
    ```
    addons:
    - name: metallb
      kind: "manifest"
      enabled: true
      namespace: boundless-system
      manifest:
        url: "https://raw.githubusercontent.com/metallb/metallb/v0.13.10/config/manifests/metallb-native.yaml"
    ```
2. Check if metallb objects are created successfully
    ```
    kubectl get crds | grep metallb
    addresspools.metallb.io             2024-01-12T16:26:30Z
    bfdprofiles.metallb.io              2024-01-12T16:26:30Z
    bgpadvertisements.metallb.io        2024-01-12T16:26:30Z
    bgppeers.metallb.io                 2024-01-12T16:26:30Z
    communities.metallb.io              2024-01-12T16:26:30Z
    ipaddresspools.metallb.io           2024-01-12T16:26:30Z
    l2advertisements.metallb.io         2024-01-12T16:26:31Z
    ```
    ```
    kubectl get all -n metallb-system
    NAME                              READY   STATUS    RESTARTS   AGE
    pod/controller-595f88d88f-tct7n   1/1     Running   0          4m44s
    pod/speaker-sm6dj                 1/1     Running   0          4m44s
    
    NAME                      TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
    service/webhook-service   ClusterIP   10.96.232.164   <none>        443/TCP   4m44s
    
    NAME                     DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
    daemonset.apps/speaker   1         1         1       1            1           kubernetes.io/os=linux   4m44s
    
    NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
    deployment.apps/controller   1/1     1            1           4m44s
    
    NAME                                    DESIRED   CURRENT   READY   AGE
    replicaset.apps/controller-595f88d88f   1         1         1       4m44s
    ```
    ```
    kubectl describe deploy -n metallb-system | grep Image
    Image:       quay.io/metallb/controller:v0.13.10
    ```
### TEST SCENARIO 2: Update add-on metallb with version v0.13.12
1. Update blueprint with new metallb version: 
    ```
    addons:
    - name: metallb
      kind: "manifest"
      enabled: true
      namespace: boundless-system
      manifest:
        url: "https://raw.githubusercontent.com/metallb/metallb/v0.13.12/config/manifests/metallb-native.yaml"
    ```
2. Check if metallb objects are created successfully
    ```
    kubectl describe deploy -n metallb-system | grep Image
    Image:       quay.io/metallb/controller:v0.13.12
    ```

### TEST SCENARIO 3: Delete manifest add-on metallb
1. Update blueprint and remove the metallb manifest addon
2. Check if addon is correctly deleted
    ```
    kubectl get addons -A
    No resources found
    ``` 
3. Check if add metallb related objects/crds are removed
    ```
    kubectl get crds | grep metallb
    ```
    ```
    kubectl get all -n metallb-system
    No resources found in metallb-system namespace.
    ```
